### ISSUE TITLE
Generators require extra-long lifetime for return statement

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#44200]
When compiling:

```rust
#![feature(generators)]

fn main() {
    let _a = || {
        if false { yield }
        let a = String::new();
        a.len()
    };
}
```

I currently get:

```
error[E0597]: `a` does not live long enough
 --> foo.rs:8:5
  |
7 |         a.len()
  |         - borrow occurs here
8 |     };
  |     ^ `a` dropped here while still borrowed
9 | }
  | - borrowed value needs to live until here

error: aborting due to previous error
```

Would be nice to not have to bind it to a local!

**Comment by Zoxc** [comment]
They way interior types are being calculated is a bit crude. It just walks all the nodes and consider them live for the scope temporaries would be. `a.len()` contains an implicit borrow of `a` and being the last statement of a block it inherits the scope for temporaries from the parent, and that scope contains a `yield`. This is not really a bug, but we'd like to find a better way to calculate these types.

**Comment by Eroc33** [comment]
@Zoxc 
Is this likely the same thing going on here? https://play.rust-lang.org/?gist=e472b9056ecc590837db32b17108bfd3&version=nightly

**Comment by Arnavion** [comment]
@Eroc33 Yes, see #44197

**Comment by alex** [comment]
This appears to have been fixed by #44392
